cdap-tms/src/main/java/co/cask/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -227,21 +235,37 @@ private long fetchAndProcessMessages() throws TopicNotFoundException, IOExceptio
 
     startTime = System.currentTimeMillis();
 
-    // Process the notifications and record the message id of where the processing is up to.
-    // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
-    long timeBoundMillis = 900L * txTimeoutSeconds;
-    MessageTrackingIterator iterator = Transactionals.execute(getTransactional(), context -> {
-      TimeBoundIterator<Message> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(), timeBoundMillis);
-      MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
-      processMessages(context, trackingIterator);
-      String lastMessageId = trackingIterator.getLastMessageId();
-
-      // Persist the message id of the last message being consumed from the iterator
-      if (lastMessageId != null) {
-        storeMessageId(context, lastMessageId);
+    MessageTrackingIterator iterator;
+    while (true) {
+      try {
+        // Process the notifications and record the message id of where the processing is up to.
+        // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
+        long timeBoundMillis = 900L * curTxTimeout;
+        iterator = Transactionals.execute(getTransactional(), curTxTimeout, context -> {
+          TimeBoundIterator<Message> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(), timeBoundMillis);
+          MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
+          processMessages(context, trackingIterator);
+          String lastMessageId = trackingIterator.getLastMessageId();
+
+          // Persist the message id of the last message being consumed from the iterator
+          if (lastMessageId != null) {
+            storeMessageId(context, lastMessageId);
+          }
+          return trackingIterator;
+        });
+        break;
+      } catch (Exception e) {
+        // we double the tx timeout if we see a tx timeout, stop doubling if it exceeds our max tx timeout
+        if (e.getCause() instanceof TransactionNotInProgressException) {
+          curTxTimeout *= 2;
+          if (curTxTimeout > maxTxTmeoutSeconds) {
+            throw e;
+          }
+        } else {
```

Comment:
don't need this else, can just throw e

Reply:
This else is needed, otherwise after we double the tx timeout, it will throw the exception